source/2018-09-18-the-git-fixup-workflow.html.markdown

```diff
@@ -0,0 +1,213 @@
+---
+title: The git fixup workflow
+date: 2018-09-18 06:58 UTC
+author: lex
+tags:
+---
+
+## Introduction
+
+The [GitHub flow](https://guides.github.com/introduction/flow/) or a similar git
+strategy where the main work is done on feature branches which result in pull
+or merge requests became for many developer teams the standard way to work with
+git and GitHub or GitLab (or Bitbucket even). Besides the main advantages of
+this way of working it also enables developers to keep a clean git history (see
+also [5 rules to Git
+better](https://blog.panter.ch/2016/08/31/5-rules-to-git-better.html)).
+
+Let's assume the following as a starting point for the suggested workflow in
+this post:
+
+- You want to maintain a clean git history, meaning that fix commits should be
+  avoided
+- You don't want to interrupt your development workflow by having to rewrite
+  your git history all the time
+- You can use `git commit --amend` only for the latest commit 
+
+Git has some powerful commands that allow us to both keep a clean history and
+not having to interrupt our development workflow. We'll call it "the git fixup
+workflow".
+
+
+## The linear worflow
+
+To illustrate the git fixup workflow we'll use the following example hack
+session:
+
+```bash
+# coding the user component's index view
+
+$ git commit -m "add user component's index view"
+
+# coding the user component's edit view
+
+$ git commit -m "add user component's edit view"
+
+# updating the component's index view so it actually works
+
+$ git commit -m "add user component's index view, it works now"
+
+# a typo is noticed in the index view
+
+$ git commit -m "fix typo in index view"
+
+# fix a bug in the view rendering
+
+$ git commit -m "fix view rendering"
+
+$ rubocop
+$ yarn lint
+
+# fix linting errors
+
+$ git commit -m "fix linting errors"
+
+# team members do a code review on the PR / MR
+
+$ git commit -m "fix import in index"
+```
+
+The above session leads to the following git history:
+
+```
+fix import in index
+fix linting errors
+fix view rendering
+fix typo in index view
+add user component's index view, it works now
+add user component's edit view
+add user component's index view
+```
+
+The following commits reflect the actual work that's relevant from another
+developer's standpoint:
```

Comment:
Maybe "a reviewer's" instead of "another developer's"

Reply:
It's not just about the review process, but about the project as a whole.

Reply:
Ok fair enough :)